### PR TITLE
Versioned block support

### LINKED ISSUE #140: Beacon chain fork support
We currently have hardcoded bellatrix blocks and other beacon objects, we should adopt the versioned structs found in eth2-client with switch statements in spec to support future beacon forks

### PR DESCRIPTION
**Pull request by olegshmuelov**
- Versioned block support for regular/blinded beacon block
- Deprecate `Bellatrix` ===> Introduce `Capella`
- Update tests with new roots
closes #140 

THIS PR split to 2 different PRs:
 - basic support for versioned block -  https://github.com/bloxapp/ssv-spec/pull/208
 - TBD. deprecates older forks & introduce Capella with state comparison.

### CODE REVIEW

**Comment by olegshmuelov** [comment]
Can be merged only after https://github.com/attestantio/go-eth2-client/issues/50 resolved

File: go.mod
Comment (alonmuroch) [inline go.mod:74]: needs to be replaced to be merged

File: types/ssz.go
Comment (alonmuroch) [inline types/ssz.go:86]: spec tests

File: types/testingutils/ssv_msgs.go
Comment (alonmuroch) [inline types/testingutils/ssv_msgs.go:311]: add the new cappella fields like withdrawals

**Review by alonmuroch — CHANGES_REQUESTED** [review]

File: types/spectest/tests/consensusdata/invalid_bellatrix_blinded_block_validation.go
Comment (olegshmuelov) [reply types/spectest/tests/consensusdata/invalid_bellatrix_blinded_block_validation.go:3]: added spectests under `types/consensusdata`

File: types/testingutils/beacon_node.go
Comment (GalRogozinski) [inline types/testingutils/beacon_node.go:527]: should we also check if the sig is nil?
not sure

File: types/testingutils/beacon_node.go
Comment (GalRogozinski) [inline types/testingutils/beacon_node.go:554]: ditto sig nil check?

File: types/generate.go
Comment (GalRogozinski) [inline types/generate.go:19]: :+1:

File: types/ssz.go
Comment (GalRogozinski) [inline types/ssz.go:89]: is there a reason why we don't use a pointer here?

File: types/ssz.go
Comment (GalRogozinski) [inline types/ssz.go:93]: ditto

File: types/ssz.go
Comment (GalRogozinski) [inline types/ssz.go:110]: ditto

**Review by GalRogozinski — CHANGES_REQUESTED** [review]

File: types/spectest/tests/consensusdata/proposer/versioned_block_validation.go
Comment (olegshmuelov) [reply types/spectest/tests/consensusdata/proposer/versioned_block_validation.go:13]: ditto

File: types/spectest/tests/ssz/test.go
Comment (olegshmuelov) [reply types/spectest/tests/ssz/test.go:38]: done

Comment (olegshmuelov) [reply types/spectest/tests/ssz/test.go:38]: duplicate

File: types/spectest/tests/consensusdata/proposer/versioned_blinded_block_unknown_version.go
Comment (olegshmuelov) [reply types/spectest/tests/consensusdata/proposer/versioned_blinded_block_unknown_version.go:12]: done

File: types/spectest/tests/ssz/withdrawals_marshaling_validation.go
Comment (olegshmuelov) [reply types/spectest/tests/ssz/withdrawals_marshaling_validation.go:20]: This way we can see the whole picture

File: types/spectest/tests/consensusdata/proposer/versioned_block_spectest.go
Comment (olegshmuelov) [reply types/spectest/tests/consensusdata/proposer/versioned_block_spectest.go:64]: done

File: types/testingutils/beacon_node.go
Comment (olegshmuelov) [reply types/testingutils/beacon_node.go:527]: not needed will fail validation

File: types/testingutils/beacon_node.go
Comment (olegshmuelov) [reply types/testingutils/beacon_node.go:554]: ditto

File: types/ssz.go
Comment (olegshmuelov) [reply types/ssz.go:89]: lets derive a task for this

File: types/ssz.go
Comment (olegshmuelov) [reply types/ssz.go:93]: ditto

File: types/spectest/tests/consensusdata/proposer/versioned_blinded_block_validation.go
Comment (GalRogozinski) [reply types/spectest/tests/consensusdata/proposer/versioned_blinded_block_validation.go:7]: I think what we can do is have SpecTests that test the `HashTreeRoot()`for different objects.
Those tests will have the expected results hardcoded.

Then we can assume that no need to hardcode expected roots for the rest of the tests.

File: types/consensus_data.go
Comment (y0sher) [inline types/consensus_data.go:151]: if we're basically deprecating Bellatrix, this is a preparation to support future version ?

Comment (olegshmuelov) [reply types/consensus_data.go:151]: exactly, so if for example next fork [`Deneb`](https://crypto.news/days-after-shanghai-ethereum-developers-readying-for-key-eip-and-deneb-upgrade/) will introduce new block, we just need to upgrade the eth2-go-client package and adjust the code to use the `Deneb` beacon block

File: types/testingutils/beacon_node.go
Comment (y0sher) [inline types/testingutils/beacon_node.go:85]: don't we have `capella.Transaction`? or this is a mistake?

**Review by y0sher — APPROVED** [review]

File: types/testingutils/beacon_node.go
Comment (olegshmuelov) [reply types/testingutils/beacon_node.go:85]: no, the transactions introduced in Bellatrix fork

File: types/spectest/tests/consensusdata/proposer/versioned_blinded_block_validation.go
Comment (GalRogozinski) [reply types/spectest/tests/consensusdata/proposer/versioned_blinded_block_validation.go:7]: https://bloxxx.atlassian.net/browse/IO1-3182

**Review by GalRogozinski — APPROVED** [review]

File: ssv/proposer.go
Comment (alonmuroch) [inline ssv/proposer.go:193]: not sure how significant it is but versioned block is "wasteful" since ssz will fill in zero data in some fields

**Review by alonmuroch — CHANGES_REQUESTED** [review]

**Comment by alonmuroch** [comment]
@olegshmuelov @y0sher @GalRogozinski @lior-blox all the test root changes, anyone went one by one to see they actually are correct? A lot was changed in this PR so it needs to be checked one-by-one.


@olegshmuelov do we have a test there actually testing different versions of the block? seems we just changed everything to cappella but no test for actually multi version support

**Comment by y0sher** [comment]
@alonmuroch  wouldn't the tests not pass if they aren't correct? I believe @olegshmuelov went one by one and change them. one of us can double check if that is needed and it can be wrong.

@olegshmuelov about the second point I agree we should have a test that covers the case where we got a block with a different version than Cappella.

**Comment by alonmuroch** [comment]
> @alonmuroch wouldn't the tests not pass if they aren't correct? I believe @olegshmuelov went one by one and change them. one of us can double check if that is needed and it can be wrong.
The tests pass but we don't know against which post root as we changed them. If the post root is not correct it will pass anyhow. That's why we introduced state comparison but it's not yet implemented here.
I suggest someone to go root by root with breakpoints to see it actually produces the correct state

**Comment by GalRogozinski** [comment]
Let's go over one by one again when we add state comparisons @alonmuroch

**Comment by GalRogozinski** [comment]
@olegshmuelov 
After discussing with Alon and @lior-blox, it seems that the best way forward is to split the PR this way:
1. Add versioned block support but don't upgrade to Capella.
2. Upgrade to Capella with state compariosn feature. This way we ensure the state rrots are correct and we add more state comparison tests.

File: go.mod
Comment (olegshmuelov) [reply go.mod]: The PR was ready before the `go-eth2-client` tag was released
https://github.com/bloxapp/ssv-spec/pull/171#issuecomment-1478337320

**Comment by olegshmuelov** [comment]
> @olegshmuelov @y0sher @GalRogozinski @lior-blox all the test root changes, anyone went one by one to see they actually are correct? A lot was changed in this PR so it needs to be checked one-by-one.
> 
> @olegshmuelov do we have a test there actually testing different versions of the block? seems we just changed everything to cappella but no test for actually multi version support

1. Yes it was checked by me, also we will probably add state comparison tests to this PR as well before merging it
2. First this PR was introduced with all block versions for regular/blinded blocks but then was requested by you to remove it. 
This why we deprecated all Bellatrix block tests and switched to use Capella
relevant [commit](https://github.com/bloxapp/ssv-spec/pull/171/commits/d021b1e5782bc2d8d48122588dcf80c246be92df#diff-c02f04705d9ade24a834f46564e331fbeef4cf8fa826c4a3071b0f946f177fe1):

File: ssv/proposer.go
Comment (olegshmuelov) [reply ssv/proposer.go:195]: If I understand your comment correctly, it doesn't make sense because we are only marshaling SSZ for specific blocks like Bellatrix, Capella, etc., rather than the entire VersionedBlock

**Comment by GalRogozinski** [comment]
Will be split to several different PRs

**Pull request closed without merging**
